doc/owl.1
.TH owl 1 "21 May 2002"
.SH NAME
owl \- tty based zephyr client
.SH SYNOPSIS
.B owl
[ \-n
]
[ \-c
.I configfile
]
[ \-t
.I tty
]
[ \-v
]
[ \-d
]
[ \-D
]
[ \-h
]

.br
.SH DESCRIPTION
.B Owl
is a fully integrated tty based instant messaging client.  Currently
it supports AOL Instant Messenger and MIT Zephyr.  It is curses-based,
allows for emacs style editing of outgoing messages and uses a perl
configuration language for setting options and customizing message
formatting.  Owl will also run happily without a configuration file.

Once Owl is started, typing 'h' will display a help screen.  Typing
':' enters command mode, allowing the user to type an owl command
line.

.PP
.SH USE
The following command line options are avilable when running owl:

.B \-n
.IP
Do not subscribe to zephyr messages on startup.  By default Owl
subscribes to the default subscriptions and to anything found in
~/.zephyr.subs.  When this option is used no subscriptions are loaded.
.LP

.B \-c \fIconfigfile\fP
.IP
Specifiy an alternate config file for Owl to use.  The default
configuration file is ~/.owlconf
.LP

.B \-t \fItty\fP
.IP
Specifiy the tty name to use for the zephyr location.
.LP

.B \-v
.IP
Print the version number of owl and exit.
.LP

.B \-d
.IP
Enable debugging.  By default debugging information is placed in
/var/tmp/owldebug.
.LP

.B \-D
.IP
Enable debugging, but first delete any existing debugging file.
.LP

.B \-h
.IP
Print command line option help.
.LP

.SH AUTHOR
Written by James Kretchmar at the Massachusetts Institute of
Technology.
Comments, questions, and bug reports may be mailed to
\fBbug-ktools@mit.edu\fP.


NEWS

$Id$

What's new in owl version 1.1
-----------------------------
* Generic filtering now exists

* The filter command creates a filter

* The view command narrows the view to a filter

* The viewclass / vc command automatically narrows the view
  to a particular class

* The viewuser / vu command automatically narrows the view
  to a particular user

* M-n automatically narrows the view to whatever is appropriate for
  the current message

* v will start a view command

* X will expunge and change the view to the view_home variable value

* R will reply to sender

* C-r and C-R will let the user edit the reply line

* T now uses the 'trash' filter

* T tells you how many messages were marked for deletion

* The local realm has been removed from login/logout messages

* Command history works

* Prettier status command

* Leave the pointer near the current message after an expunge

* C-l recenters the typing window

* zlocate is implemented

* @italic works like @i

* Unecessary quoting and options in reply lines have been removed

* Errors in the 'info' command don't show up in a popup now

* Implemented the aexec, pexec commands

* The zsig now gets zephyr formatting

* Messages now have message id's. They are displayed in 'info' and are
  passed to the config as $owl::id.

* The 'show' command is implemented to display info about filters.

* M-D marks for deletion all messages in the current view

* added the delete command

* load-subs with no argument loads the default subscription file

* "Truncated" is now displayed when the *current* message is truncated
  not when the last displayed message is truncated.

* The reply-lockout filter locks out replies to some classes

* In the configfile owl::receive_msg is run when a message is
  received.

* Added the beep command

* Added the contributors file

* zaway, zaway_msg and zaway_msg_default are all user variables now

* There is a zsig variable that overrides other zsig mechanisims

* The appendtosepbar variable exists to allow the config to display
  things in the sepbar

* login messages are included as 'trash' messages by default now

* Fixed a memory corruption bug in readconfig.c

* There are example configs in the distribution now

* Implemented the perl, aperl and pperl commands

* FIxed bug with zsigproc zombies

* Can disable zsigproc after setting it now

* There is support for multiple browsers with the 'webbrowser'
  variable.

* Set now takes a -q option to not echo

* debug is now a variable that can be turned on and off

* Don't allow " and \ characters in URLs passed to a webbrowser

* Build system cleaned up some

* Pop up windows are wider by default

* Fix to work with older versions of libzephyr

* PageUp and PageDown work in the main window and in popups

* Added doc/code.txt describing the code layout


What's new in owl version 1.0
-----------------------------
* 'print' and 'set' with no arguments prints all variables

* The 'unsub' command renamed to 'unsuball'

* 'unsubscribe' (and 'unsub') command

* 'getsubs' command to print subscriptions

* Bug fixed in logging messages to more than one recipient

* Support for zwrite options -C, -O and -n

* Fixed owl_editwin_delete_char bug when there are no chars after the
  cursor

* Display 'more' and 'truncated' in the status bar

* Enable printing of the 'zsigproc' and 'loginsubs' variables

* Only allow message scrolling if the message is longer than the
  screen

* 'P' (capitolized) will advance the pointer to the next personal
  message.  'M-P' will move the pointer to the previous personal
  message.

* 'T' will mark certain "trash" messages for deletion.  This includes:
  mail notifications, pings and admin messages.

* Replying to a login/logout message sends as a personal message to
  the user

* Replying to an admin message that's an outgoing message copy sends
  as a personal message to the user

* 'status' command

* intro document added to the release

* The local realm is trimmed everywhere, including in logging

* Fixed off by one bug in scrolling popups

* Complete on line help added

* Pass the realm to the config file in $owl::realm

* Word wrap and scroll at the bottom right corner of the editing
  window

* Fixed a bug in editwing window key_right past the last character

* Print an error and don't start owl if the configfile can't be
  parsed

* Fix bug where the only message displayed might not be the current
  message

* Don't display the M-^Z key error on resize

What's new in owl version 0.11
------------------------------
* Bug fixed in viewing messages longer than a screen

* Indicate in the sepbar if there is a vertical offset present

* '.' on a line by itself will send a zephyr

* disable-ctrl-d variable will disable sending on C-d

* bug fixed that kept C-k from deleting a final newline

* make non-character meta keys work

* fixed phatom zephyr problem

* <, >, M-< and M-> work in popups now

* removed extra spaces at bottom of a popup

* Added "about" command

* Fixed bug using 'M' with no current message

* Log personal messages to 'all' as well as the personal file

* Messages take up less memory

What's new in owl version 0.10
------------------------------
* Scrollable / pageable popup windows.  They act like the 'less' or
  'more' programs.

* Scrolling within a message, to facilitate reading messages longer
  than the screen size.

* The command parser deals with quoting.  i.e. zwrite -c "foo class"
  now works.

* The slow cut and paste bug is fixed.

* C-v and M-v will page down and up in the main window, respectively.

* A bug in redrawing the popup windows is fixed.

* An error is displayed when non-functional keys are pressed.

* @i() formatting is displayed as underlined in curses.

* Curly braces work for zephyr formatting.

* The popup window size is relatively larger on small terminals

* A bug in having meta keys work on eight bit input xterms is fixed.

* The cursor location is more consistent.



owl.c
/*
 * Copyright 2001 Massachusetts Institute of Technology
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby
 * granted, provided that the above copyright notice and this
 * permission notice appear in all copies and in supporting
 * documentation.  No representation is made about the suitability of
 * this software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <time.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "owl.h"

static const char fileIdent[] = "$Id$";

owl_global g;

int main(int argc, char **argv, char **env) {
  WINDOW *recwin, *sepwin, *typwin, *msgwin;
  owl_editwin *tw;
  owl_popwin *pw;
  int j, ret, initialsubs, debug, argcsave, followlast;
  int newmsgs, zpendcount, nexttimediff;
  struct sigaction sigact;
  char *configfile, *tty, *perlout, *perlerr, **argvsave, buff[LINE], startupmsg[LINE];
  owl_filter *f;
  owl_style *s;
  time_t nexttime, now;
  struct tm *today;
  char *dir;
#ifdef HAVE_LIBZEPHYR
  ZNotice_t notice;
#endif

  argcsave=argc;
  argvsave=argv;
  configfile=NULL;
  tty=NULL;
  debug=0;
  initialsubs=1;
  if (argc>0) {
    argv++;
    argc--;
  }
  while (argc>0) {
    if (!strcmp(argv[0], "-n")) {
      initialsubs=0;
      argv++;
      argc--;
    } else if (!strcmp(argv[0], "-c")) {
      if (argc<2) {
	fprintf(stderr, "Too few arguments to -c\n");
	usage();
	exit(1);
      }
      configfile=argv[1];
      argv+=2;
      argc-=2;
    } else if (!strcmp(argv[0], "-t")) {
      if (argc<2) {
	fprintf(stderr, "Too few arguments to -t\n");
	usage();
	exit(1);
      }
      tty=argv[1];
      argv+=2;
      argc-=2;
    } else if (!strcmp(argv[0], "-d")) {
      debug=1;
      argv++;
      argc--;
    } else if (!strcmp(argv[0], "-v")) {
      printf("This is owl version %s\n", OWL_VERSION_STRING);
      exit(0);
    } else {
      fprintf(stderr, "Uknown argument\n");
      usage();	      
      exit(1);
    }
  }

#ifdef HAVE_LIBZEPHYR
  /* zephyr init */
  ret=owl_zephyr_initialize();
  if (ret) {
    exit(1);
  }
#endif
  
  /* signal handler */
  sigact.sa_handler=sig_handler;
  sigemptyset(&sigact.sa_mask);
  sigact.sa_flags=0;
  sigaction(SIGWINCH, &sigact, NULL);
  sigaction(SIGALRM, &sigact, NULL);

  /* screen init */
  sprintf(buff, "TERMINFO=%s", TERMINFO);
  putenv(buff);
  initscr();
  start_color();
#ifdef HAVE_USE_DEFAULT_COLORS
  use_default_colors();
#endif
  raw();
  noecho();

  /* define simple color pairs */
  if (has_colors() && COLOR_PAIRS>=8) {
    init_pair(OWL_COLOR_BLACK,   COLOR_BLACK,   -1);
    init_pair(OWL_COLOR_RED,     COLOR_RED,     -1);
    init_pair(OWL_COLOR_GREEN,   COLOR_GREEN,   -1);
    init_pair(OWL_COLOR_YELLOW,  COLOR_YELLOW,  -1);
    init_pair(OWL_COLOR_BLUE,    COLOR_BLUE,    -1);
    init_pair(OWL_COLOR_MAGENTA, COLOR_MAGENTA, -1);
    init_pair(OWL_COLOR_CYAN,    COLOR_CYAN,    -1);
    init_pair(OWL_COLOR_WHITE,   COLOR_WHITE,   -1);
  }

  /* owl global init */
  owl_global_init(&g);
  if (debug) owl_global_set_debug_on(&g);
  owl_global_set_startupargs(&g, argcsave, argvsave);
#ifdef HAVE_LIBZEPHYR
  owl_global_set_havezephyr(&g);
#endif
  owl_global_set_haveaim(&g);
  
  /* create the owl directory, in case it does not exist */
  dir=owl_sprintf("%s/%s", owl_global_get_homedir(&g), OWL_CONFIG_DIR);
  mkdir(dir, S_IRWXU);
  owl_free(dir);

  /* set the tty, either from the command line, or by figuring it out */
  if (tty) {
    owl_global_set_tty(&g, tty);
  } else {
    owl_global_set_tty(&g, owl_util_get_default_tty());
  }

  /* setup the built-in styles */
  s=owl_malloc(sizeof(owl_style));
  owl_style_create_internal(s, "default", &owl_stylefunc_default,
			    "Default message formatting");
  owl_global_add_style(&g, s);

  s=owl_malloc(sizeof(owl_style));
  owl_style_create_internal(s, "basic", &owl_stylefunc_basic,
			    "Basic message formatting.");
  owl_global_add_style(&g, s);

  s=owl_malloc(sizeof(owl_style));
  owl_style_create_internal(s, "oneline", &owl_stylefunc_oneline,
			    "Formats for one-line-per-message");
  owl_global_add_style(&g, s);

  /* setup the default filters */
  /* the personal filter will need to change again when AIM chat's are
   *  included.  Also, there should be an %aimme% */
  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "personal", "( type ^zephyr$ "
			     "and class ^message$ and instance ^personal$ "
			     "and ( recipient ^%me%$ or sender ^%me%$ ) ) "
			     "or ( type ^aim$ and login ^none$ )");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "trash", "class ^mail$ or opcode ^ping$ or type ^admin$ or ( not login ^none$ )");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "ping", "opcode ^ping$");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "auto", "opcode ^auto$");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "login", "not login ^none$");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "reply-lockout", "class ^noc or class ^mail$");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "out", "direction ^out$");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "aim", "type ^aim$");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "zephyr", "type ^zephyr$");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "none", "false");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  f=owl_malloc(sizeof(owl_filter));
  owl_filter_init_fromstring(f, "all", "true");
  owl_list_append_element(owl_global_get_filterlist(&g), f);

  /* set the current view */
  owl_view_create(owl_global_get_current_view(&g), "main", f, owl_global_get_style_by_name(&g, "default"));

  /* AIM init */
  owl_aim_init();

  /* process the startup file */
  owl_function_source(NULL);

  /* read the config file */
  owl_context_set_readconfig(owl_global_get_context(&g));
  perlerr=owl_perlconfig_readconfig(configfile);
  if (perlerr) {
    endwin();
    fprintf(stderr, "\nError parsing configfile: %s\n", perlerr);
    exit(1);
  }

  /* if the config defines a formatting function, add 'perl' as a style */
  if (owl_global_is_config_format(&g)) {
    owl_function_debugmsg("Found perl formatting");
    s=owl_malloc(sizeof(owl_style));
    owl_style_create_perl(s, "perl", "owl::_format_msg_legacy_wrap",
			  "User-defined perl style that calls owl::format_msg"
			  "with legacy global variable support");
    owl_global_add_style(&g, s);
    owl_global_set_default_style(&g, "perl");
  }

  /* set the startup and default style, based on userclue and presence of a
   * formatting function */
  if (owl_global_is_config_format(&g)) {
    owl_global_set_default_style(&g, "perl");
  } else if (owl_global_is_userclue(&g, OWL_USERCLUE_CLASSES)) {
    owl_global_set_default_style(&g, "default");
  } else {
    owl_global_set_default_style(&g, "basic");
  }

  /* execute the startup function in the configfile */
  perlout = owl_perlconfig_execute("owl::startup();");
  if (perlout) owl_free(perlout);
  owl_function_debugmsg("-- Owl Startup --");
  
  /* hold on to the window names for convenience */
  msgwin=owl_global_get_curs_msgwin(&g);
  recwin=owl_global_get_curs_recwin(&g);
  sepwin=owl_global_get_curs_sepwin(&g);
  typwin=owl_global_get_curs_typwin(&g);
  tw=owl_global_get_typwin(&g);

  wrefresh(sepwin);

  /* load zephyr subs */
  if (initialsubs) {
    /* load normal subscriptions */
    ret=owl_zephyr_loadsubs(NULL);
    if (ret!=-1) {
      owl_global_add_userclue(&g, OWL_USERCLUE_CLASSES);
    }

    /* load login subscriptions */
    if (owl_global_is_loginsubs(&g)) {
      owl_function_loadloginsubs(NULL);
    }
  }

  /* zlog in if we need to */
  if (owl_global_is_startuplogin(&g)) {
    owl_zephyr_zlog_in();
  }

  owl_view_set_style(owl_global_get_current_view(&g), 
		     owl_global_get_style_by_name(&g, owl_global_get_default_style(&g)));   

  /* welcome message */
  strcpy(startupmsg, "-------------------------------------------------------------------------\n");
  sprintf(buff,      "Welcome to Owl version %s.  Press 'h' for on-line help. \n", OWL_VERSION_STRING);
  strcat(startupmsg, buff);
  strcat(startupmsg, "                                                                         \n");
  strcat(startupmsg, "If you would like to receive release announcements about owl you can join \n");
  strcat(startupmsg, "the owl-users@mit.edu mailing list.  MIT users can add themselves,       \n");
  strcat(startupmsg, "otherwise send a request to owner-owl-users@mit.edu.               ^ ^   \n");
  strcat(startupmsg, "                                                                   OvO   \n");
  strcat(startupmsg, "Please report any bugs or suggestions to bug-owl@mit.edu          (   )  \n");
  strcat(startupmsg, "-------------------------------------------------------------------m-m---\n");
  owl_function_adminmsg("", startupmsg);
  sepbar(NULL);
  
  owl_context_set_interactive(owl_global_get_context(&g));

  nexttimediff=10;
  nexttime=time(NULL);

  /* main loop */
  while (1) {

    /* if a resize has been scheduled, deal with it */
    owl_global_resize(&g, 0, 0);

    /* these are here in case a resize changes the windows */
    msgwin=owl_global_get_curs_msgwin(&g);
    recwin=owl_global_get_curs_recwin(&g);
    sepwin=owl_global_get_curs_sepwin(&g);
    typwin=owl_global_get_curs_typwin(&g);

    followlast=owl_global_should_followlast(&g);
    
    /* Do AIM stuff */
    if (owl_global_is_doaimevents(&g)) {
      owl_aim_process_events();

      if (owl_global_is_aimloggedin(&g)) {
	if (owl_timer_is_expired(owl_global_get_aim_buddyinfo_timer(&g))) {
	  owl_buddylist_request_idletimes(owl_global_get_buddylist(&g));
	  owl_timer_reset(owl_global_get_aim_buddyinfo_timer(&g));
	}
      }
    }

    /* little hack */
    now=time(NULL);
    today=localtime(&now);
    if (today->tm_mon==9 && today->tm_mday==31 && owl_global_get_runtime(&g)<600) {
      if (time(NULL)>nexttime) {
	if (nexttimediff==1) {
	  nexttimediff=10;
	} else {
	  nexttimediff=1;
	}
	nexttime+=nexttimediff;
	owl_hack_animate();
      }
    }

    /* Grab incoming messages. */
    newmsgs=0;
    zpendcount=0;
    while(owl_zephyr_zpending() || owl_global_messagequeue_pending(&g)) {
#ifdef HAVE_LIBZEPHYR
      struct sockaddr_in from;
#endif
      owl_message *m=NULL;
      owl_filter *f;

      /* grab the new message, stick it in 'm' */
      if (owl_zephyr_zpending()) {
#ifdef HAVE_LIBZEPHYR
	/* grab a zephyr notice, but if we've done 20 without stopping,
	   take a break to process keystrokes etc. */
	if (zpendcount>20) break;
	ZReceiveNotice(&notice, &from);
	zpendcount++;
	
	/* is this an ack from a zephyr we sent? */
	if (owl_zephyr_notice_is_ack(&notice)) {
	  owl_zephyr_handle_ack(&notice);
	  continue;
	}
	
	/* if it's a ping and we're not viewing pings then skip it */
	if (!owl_global_is_rxping(&g) && !strcasecmp(notice.z_opcode, "ping")) {
	  continue;
	}

	/* create the new message */
	m=owl_malloc(sizeof(owl_message));
	owl_message_create_from_znotice(m, &notice);
#endif
      } else if (owl_global_messagequeue_pending(&g)) {
	m=owl_global_messageuque_popmsg(&g);
      }
      
      /* if this message it on the puntlist, nuke it and continue */
      if (owl_global_message_is_puntable(&g, m)) {
	owl_message_free(m);
	continue;
      }

      /* otherwise add it to the global list */
      owl_messagelist_append_element(owl_global_get_msglist(&g), m);
      newmsgs=1;

      /* let the config know the new message has been received */
      owl_perlconfig_getmsg(m, 0, NULL);

      /* add it to any necessary views; right now there's only the current view */
      owl_view_consider_message(owl_global_get_current_view(&g), m);

      /* do we need to autoreply? */
      if (owl_message_is_type_zephyr(m) && owl_global_is_zaway(&g)) {
	owl_zephyr_zaway(m);
      }

      /* ring the bell if it's a personal */
      if (owl_global_is_personalbell(&g) &&
	  !owl_message_is_loginout(m) &&
	  !owl_message_is_mail(m) &&
	  owl_message_is_private(m)) {
	owl_function_beep();
      }

      /* if it matches the alert filter, do the alert action */
      f=owl_global_get_filter(&g, owl_global_get_alert_filter(&g));
      if (f && owl_filter_message_match(f, m)) {
	owl_function_command(owl_global_get_alert_action(&g));
      }

      /* check for burning ears message */
      /* this is an unsupported feature */
      if (owl_global_is_burningears(&g) && owl_message_is_burningears(m)) {
	char *buff;
	buff = owl_sprintf("@i(Burning ears message on class %s)", owl_message_get_class(m));
	owl_function_adminmsg(buff, "");
	owl_free(buff);
	owl_function_beep();
      }

      /* log the message if we need to */
      if (owl_global_is_logging(&g) || owl_global_is_classlogging(&g)) {
	owl_log_incoming(m);
      }
    }

    /* follow the last message if we're supposed to */
    if (newmsgs && followlast) {
      owl_function_lastmsg_noredisplay();
    }

    /* do the newmsgproc thing */
    if (newmsgs) {
      owl_function_do_newmsgproc();
    }
    
    /* redisplay if necessary */
    /* this should be optimized to not run if the new messages won't be displayed */
    if (newmsgs) {
      owl_mainwin_redisplay(owl_global_get_mainwin(&g));
      sepbar(NULL);
      if (owl_popwin_is_active(owl_global_get_popwin(&g))) {
	owl_popwin_refresh(owl_global_get_popwin(&g));
	/* TODO: this is a broken kludge */
	if (owl_global_get_viewwin(&g)) {
	  owl_viewwin_redisplay(owl_global_get_viewwin(&g), 0);
	}
      }
      owl_global_set_needrefresh(&g);
    }

    /* if a popwin just came up, refresh it */
    pw=owl_global_get_popwin(&g);
    if (owl_popwin_is_active(pw) && owl_popwin_needs_first_refresh(pw)) {
      owl_popwin_refresh(pw);
      owl_popwin_no_needs_first_refresh(pw);
      owl_global_set_needrefresh(&g);
      /* TODO: this is a broken kludge */
      if (owl_global_get_viewwin(&g)) {
	owl_viewwin_redisplay(owl_global_get_viewwin(&g), 0);
      }
    }

    /* update the terminal if we need to */
    if (owl_global_is_needrefresh(&g)) {
      /* leave the cursor in the appropriate window */
      if (owl_global_is_typwin_active(&g)) {
	owl_function_set_cursor(typwin);
      } else {
	owl_function_set_cursor(sepwin);
      }
      doupdate();
      owl_global_set_noneedrefresh(&g);
    }

    /* Handle all keypresses.  If no key has been pressed, sleep for a
     * little bit, but otherwise do not.  This lets input be grabbed
     * as quickly as possbile */
    j=wgetch(typwin);
    if (j==ERR) {
      usleep(10);
      continue;
    }
    /* find and activate the current keymap.
     * TODO: this should really get fixed by activating
     * keymaps as we switch between windows... 
     */
    if (pw && owl_popwin_is_active(pw) && owl_global_get_viewwin(&g)) {
      owl_context_set_popless(owl_global_get_context(&g), 
			      owl_global_get_viewwin(&g));
      owl_function_activate_keymap("popless");
    } else if (owl_global_is_typwin_active(&g) 
	       && owl_editwin_get_style(tw)==OWL_EDITWIN_STYLE_ONELINE) {
      /*
      owl_context_set_editline(owl_global_get_context(&g), tw);
      owl_function_activate_keymap("editline");
      */
    } else if (owl_global_is_typwin_active(&g) 
	       && owl_editwin_get_style(tw)==OWL_EDITWIN_STYLE_MULTILINE) {
      owl_context_set_editmulti(owl_global_get_context(&g), tw);
      owl_function_activate_keymap("editmulti");
    } else {
      owl_context_set_recv(owl_global_get_context(&g));
      owl_function_activate_keymap("recv");
    }
    /* now actually handle the keypress */
    ret = owl_keyhandler_process(owl_global_get_keyhandler(&g), j);
    if (ret!=0 && ret!=1) {
      owl_function_makemsg("Unable to handle keypress");
    }
  }
}

void sig_handler(int sig) {
  if (sig==SIGWINCH) {
    /* we can't inturrupt a malloc here, so it just sets a flag
     * schedulding a resize for later
     */
    owl_function_resize();
  }
}

void usage() {
  fprintf(stderr, "Owl version %s\n", OWL_VERSION_STRING);
  fprintf(stderr, "Usage: owl [-n] [-d] [-v] [-h] [-c <configfile>] [-t <ttyname>]\n");
  fprintf(stderr, "  -n      don't load zephyr subscriptions\n");
  fprintf(stderr, "  -d      enable debugging\n");
  fprintf(stderr, "  -v      print the Owl version number and exit\n");
  fprintf(stderr, "  -h      print this help message\n");
  fprintf(stderr, "  -c      specify an alternate config file\n");
  fprintf(stderr, "  -t      set the tty name\n");
  
}
